### ISSUE TITLE
What is the difference between jax.lax.stop_gradient and eqx.static_field=True ?

### ISSUE THREAD

**Issue by timnotavailable** [patrick-kidger/equinox#958]
Hello,
I'm going to make a ODE simulator for Bloch Equation, I would like to define a module which has some attributes of array type,

class(eqx.Module):
array1:Array["x y z"]
array2: Array ["x y z"]

what I would like to do is : When I'm initializing with a np.array, then it is static and do not need the gradient, when initializing with a jnp.array then it is marked dynamic in jax (requires gradient). Currently I read the whole document, and there are two ways doing so : eqx.static_field = True or jax.lax.stop_gradient , which one should I choose and how should I wrap this function into the initialization?

I've also checked some issues, it is said that eqx.static_field = True this way is dangerous, what should I do to realize it?

**Comment by johannahaffner** [comment]
I think what you want here is custom filtering, e.g. as done in [this example](https://docs.kidger.site/diffrax/examples/kalman_filter/) (look for the `filter_spec` definition in the `main` function). 

Would that work for your use case? It would allow you to control what ends on in the dynamic and static partition of your PyTree, in case you are using that during optimisation.

**Comment by patrick-kidger** [comment]
You want `jax.lax.stop_gradient`. This will block autodiff from operating through that variable.

`eqx.field(static=True)` does something totally different -- it marks a dataclass field as not being part of the pytree structure.

As for how to use `jax.lax.stop_gradient`, usually something like this:
```python
class AddWithoutGradient(eqx.Module):
    x: jax.Array

    def __call__(self, y: jax.Array):
        x = jax.lax.stop_gradient(x)
        return x + y
```

**Comment by johannahaffner** [comment]
`jax.lax.stop_gradient` "gets lost" when the input to a module is wrapped, and seems to require doing the wrapping directly inside the bound methods. This is why I didn't recommend it yesterday - although we might be able to fix it though the module metaclass?

Take this example, which I would naively expect to work:

```python
import equinox as eqx
import jax
import wadler_lindig as wl

class ArrayContainer(eqx.Module):
    x: jax.Array

def parabolic_loss(tree):
    return jnp.sum(tree.x**2)

some_vector = 2.0 * jax.numpy.ones(3)

a = ArrayContainer(some_vector)
wl.pprint(jax.grad(parabolic_loss)(a), short_arrays=False)  # Has gradient

b = ArrayContainer(jax.lax.stop_gradient(some_vector))
wl.pprint(jax.grad(parabolic_loss)(b), short_arrays=False)  # Also has gradient
```

**Comment by patrick-kidger** [comment]
@johannahaffner -- so this is a misunderstanding of how `jax.lax.stop_gradient` works. It has to be called from within the region that has `jax.grad` applied to it. It doesn't control a property of the array it is called on -- it's a function that is called within a computation graph (and this function is the identity function with zero gradient).

**Comment by johannahaffner** [comment]
Thank you for clearing that up! 

FWIW, here are some collected related issues, these four are directly relevant

https://github.com/patrick-kidger/equinox/issues/909, https://github.com/patrick-kidger/equinox/issues/710 https://github.com/patrick-kidger/equinox/issues/31, including [this comment](https://github.com/patrick-kidger/equinox/issues/31#issuecomment-1036496469), https://github.com/patrick-kidger/equinox/issues/214

**Comment by patrick-kidger** [comment]
I think option 1 sounds like a good choice! Although FWIW you probably won't need to use static fields at all -- using `eqx.filter_{jit, grad, ...}` is generally a better choice. (Static fields are really an advanced feature that I try to have people avoid using unless they're familiar with the details of e.g. how JIT caching interacts with pytree semantics.)

Note that JAX totally will compile properties though. JAX uses a tracing compiler, [see point 2 here](https://kidger.site/thoughts/torch2jax/).
